ptf/saiwred.py

```diff
@@ -0,0 +1,1157 @@
+        self.assertTrue(self.wred_ecn != 0)
+        # Create dropping WRED profile (WredDropIPv4Test)
+        self.wred_drop = sai_thrift_create_wred(
+            self.client,
+            green_enable=True,
+            green_min_threshold=0,
+            green_max_threshold=max_thr,
+            green_drop_probability=100)
+        self.assertTrue(self.wred_drop != 0)
+        print("-----------------------")
+
+    def tearDown(self):
+        """
+        Deconfigure WredTest:
+
+        1. Remove wred profiles
+           - self.wred_ecn - "mark all" profile
+           - self.wred_drop - "drop all" profile
+        2. Deconfigure routing
+           - remove all created route entries
+           - remove all created neighbor entries
+           - remove all created nexthops
+        1. Call SaiHelper tearDown()
+        """
+        print("\n\n-----------------------")
+        print("Tear down WRED tests...")
+        #######################
+        # Remove WRED profiles
+        #######################
+        print("Remove WRED profiles...")
+        status = sai_thrift_remove_wred(self.client, self.wred_ecn)
+        self.assertEqual(status, SAI_STATUS_SUCCESS)
+        status = sai_thrift_remove_wred(self.client, self.wred_drop)
+        self.assertEqual(status, SAI_STATUS_SUCCESS)
+        ################
+        # Remove routes
+        ################
+        print("Deconfigure routing...")
+        for route in self.routes:
+            status = sai_thrift_remove_route_entry(self.client, route)
+            self.assertEqual(status, SAI_STATUS_SUCCESS)
+        for neighbor in self.neighbors:
+            status = sai_thrift_remove_neighbor_entry(self.client, neighbor)
+            self.assertEqual(status, SAI_STATUS_SUCCESS)
+        for nexthop in self.nexthops:
+            status = sai_thrift_remove_next_hop(self.client, nexthop)
+            self.assertEqual(status, SAI_STATUS_SUCCESS)
+
+        super(WredBaseTest, self).tearDown()
+
+    @staticmethod
+    def _statsDiff(i_stats, e_stats):
+        """ Returns a dict off stats that differ
+        Args:
+            i_stats(dict): i stats
+            e_stats(dict): e stats
+
+        Returns:
+            dect: diff
+        """
+        return {stat: e_stats[stat] - i_stats[stat]
+                for stat in e_stats
+                if e_stats[stat] - i_stats[stat]}
+
+    @staticmethod
+    def _printStats(*stats):
+        for (name, i_stats, e_stats) in stats:
+            print("\t", name, "stats:", WredBaseTest._statsDiff(i_stats,
+                                                                e_stats))
+
+    def _verifyStat(self, i_stats, e_stats, stat, delta):
+        self.assertEqual(e_stats[stat]-i_stats[stat], delta)
+
+    def _verifyStats(self, i_stats, e_stats, *stats):
+        print("Verify stats:")
+        for (name, delta) in stats:
+            print("|", name, "=>", delta)
+            self._verifyStat(i_stats, e_stats, name, delta)
+
+
```

Comment:
This test (traffic test) fails (on model), packet is not ECN-marked as expected (packet verify fails). I do not know why. As I see in jenkins file the test is excluded from the CI scope and now I am not sure if I should remove the test - exclude from the scope to OCP (at least before understanding/fixing failure behavior) or just leave it as is. What do you suggest?

Reply:
Model doesn't have TM which provides the queue_depth. Queue depth is used for WRED meters to determine if the packet can be ECN marked or dropped. Are we validating these tests in hardware switch??